### ISSUE TITLE
wx.Mask() not working on macOS

### ISSUE THREAD

**Issue by bal-agates** [wxWidgets/Phoenix#1377]
**Operating system**: macOS 10.13.6
**wxPython version & source**:  4.0.6 osx-cocoa (phoenix) wxWidgets 3.0.5 (pypi)
**Python version & source**: 3.7.4 (default, Jul 11 2019, 01:07:24) (MacPorts)

**Description of the problem**: On macOS the wx.Mask() has several problems that might be related.  When the drawing area is the same as a masked image the mask is inverted using wx.Mask(bitmap).  When the drawing area is different than a masked image with rectangular mask the mask is not as expected and includes diagonal bands (hinting at width and height differences between mask and image).  Using wx.Mask(bitmap,color) produces the expected results but is orders of magnitude slower than wx.Mask(bitmap) for large images.
![problem_false_false_50](https://user-images.githubusercontent.com/55847433/65706009-1e7c1f80-e04f-11e9-8d2a-fe393ee3ccb3.png)
![expected_true_false_50](https://user-images.githubusercontent.com/55847433/65706017-20de7980-e04f-11e9-9b41-70b50b934ccf.png)

This problem is present on a slightly earlier version on macOS 4.0.4 osx-cocoa (phoenix) wxWidgets 3.0.5.

This problem is not present on Ubuntu 18.04 4.0.1 gtk3 (phoenix).

```python
"""
wx_mask_bug1.py - test wx mask bug

I am having problems with the wx.Mask() function on macOS.  This is an attempt
to simplify the problem.  There appears to be two problems on macOS.  The same
test program on Ubuntu 18.04 with GTK behaves as expected.

1) wx.Mask(bitmap) appears to invert mask.  The documentation at wxPython.org
says "...the masked area is black and the unmasked area is white." and "...the
unmasked area of the bitmap will be drawn, and the masked area will not be
drawn."  To demonstrate this set (MASK_WITH_COLOR=False,
INVERT_MASK_COLORS=False, IMAGE_SIZE_REDUCTION=0).  The image looks as expected
using settings (MASK_WITH_COLOR=False, INVERT_MASK_COLORS=True,
IMAGE_SIZE_REDUCTION=0).

2) wx.Mask(bitmap,color) seems to work but wx.Mask(bitmap) sometimes produces
weird mask patterns when the image size and window size are different.  To
demonstrate this set (MASK_WITH_COLOR=False, INVERT_MASK_COLORS=False,
IMAGE_SIZE_REDUCTION=50).

I can workaround the problems on macOS using the settings (MASK_WITH_COLOR=True,
INVERT_MASK_COLORS=False,  IMAGE_SIZE_REDUCTION=0). My ultimate application is
to mask the map collar on some very large map images. When it works
wx.Mask(bitmap) is MUCH faster on macOS than wx.Mask(bitmap,color) for large
images.  The first takes milli-seconds where the later takes seconds.

Brett Larson
25-Sep-2019
"""
import sys
import wx

MASK_WITH_COLOR = False
INVERT_MASK_COLORS = False
IMAGE_SIZE_REDUCTION = 50

class MyFrame(wx.Frame):

    def __init__(self, parent=None, id=wx.ID_ANY, pos=(10,30), title='Mask example'):
        """ Initialize MyFrame creating masked image. """

        # Initialize super
        window_size = wx.Size(640, 480)
        wx.Frame.__init__(self, parent, id, title, pos, window_size)

        # Set the client size and bind paint event
        self.SetClientSize(window_size)
        self.Bind(wx.EVT_PAINT, self.on_paint)

        # Create an image to mask that is slightly smaller than client area.
        width = window_size.width - IMAGE_SIZE_REDUCTION
        height = window_size.height - IMAGE_SIZE_REDUCTION
        self.image_bmp = wx.Bitmap(wx.Size(width, height))
        image_dc = wx.MemoryDC(self.image_bmp)
        image_dc.SetBackground(wx.WHITE_BRUSH)
        image_dc.Clear()
        image_dc.SetPen(wx.BLACK_PEN)
        for x in range(0, width, 10):
            image_dc.DrawLine(x, 0, x, height)
        for y in range(0, height, 10):
            image_dc.DrawLine(0, y, width, y)
        #self.image_bmp.SaveFile("image.bmp", wx.BITMAP_TYPE_BMP)

        # Create a mask the same size as image.
        self.mask_bmp = wx.Bitmap(self.image_bmp.GetSize(), depth=1)
        mask_dc = wx.MemoryDC(self.mask_bmp)
        if INVERT_MASK_COLORS:
            mask_dc.SetBackground(wx.WHITE_BRUSH)
            mask_dc.SetBrush(wx.BLACK_BRUSH)
        else:
            mask_dc.SetBackground(wx.BLACK_BRUSH)
            mask_dc.SetBrush(wx.WHITE_BRUSH)
        mask_dc.Clear()
        BORDER = 50
        mask_corners = [(BORDER, BORDER), (width-BORDER, BORDER),
                        (width-BORDER, height-BORDER), (BORDER, height-BORDER)]
        mask_dc.DrawPolygon(mask_corners)
        #self.mask_bmp.SaveFile("mask.bmp", wx.BITMAP_TYPE_BMP)

        if MASK_WITH_COLOR:
            print("creating mask with color")
            mask = wx.Mask(self.mask_bmp, wx.BLACK)
        else:
            print("creating mask without color")
            mask = wx.Mask(self.mask_bmp)

        # Apply the mask to image
        self.image_bmp.SetMask(mask)

    def on_paint(self, _event):
        """ Handle EVT_PAINT events. """

        # Draw the image
        dc = wx.PaintDC(self)
        dc.SetBackground(wx.GREEN_BRUSH)
        dc.Clear()
        dc.DrawBitmap(self.image_bmp, 0, 0, True)

class MyApp(wx.App):

    def OnInit(self):
        print(sys.version)
        print(wx.version())
        self.frame = MyFrame()
        self.frame.Show()
        self.SetTopWindow(self.frame)
        return True

app = MyApp(0)
app.MainLoop()
```

**Comment by RobinD42** [comment]
It looks like this has already been fixed in wxWidgets 3.1, so it will be included in wxPython 4.1. However, I did have to remove the bitmap from the `mask_dc` to avoid an assertion, like this:

```
    mask_dc.SelectObject(wx.NullBitmap)
```

Unfortunately that did not help when running wxPython 4.0.6.

**Comment by bal-agates** [comment]
I am going to close this issue as it appears to be fixed in wxPython 4.1.0/wxWidgets 3.1.4.  With the newer wxPython I ran into an issue that illuminated another problem with my example on macOS (i.e. the default drawing mode is anti-aliased and even with depth=1 non B/W pixels were generated for diagonal lines causing wxWidgets errors).  I needed to disable the anti-aliasing.  Enclosed is an updated reference app.

```
"""
wx_mask_bug1.py - test wx mask bug

With wxPython 4.0.7.post2/wxWidgets 3.0.5 I had problems with the wx.Mask() on
macOS.  Under certain window size and image size the mask was a very different
shape than expected.

With wxPython 4.1.0/wxWidgets 4.1.4 wx.Mask() appears to be functioning as
expected on macOS, however, I needed to modify this program.  When the mask had
diagonal lines the mask image was anti-aliased and I got the following error

    wx._core.wxAssertionError: C++ assertion "v == 0 || v == 3*0xFF" failed at
    /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/osx/core/bitmap.cpp(1563)
    in InitFromMonoBitmap(): Non-monochrome bitmap supplied images.

even though I was only drawing with BLACK and WHITE and I had specified a
depth of 1 (on macOS I do not think the depth is totally respected):

"""
import sys
import wx

IMAGE_SIZE_REDUCTION = 50
MASK_BORDER = 50
MASK_SKEW = 5

class MyFrame(wx.Frame):

    def __init__(self, parent=None, id=wx.ID_ANY, pos=(10,30), title='Mask example'):
        """ Initialize MyFrame creating masked image. """

        # Initialize super
        window_size = wx.Size(640, 480)
        wx.Frame.__init__(self, parent, id, title, pos, window_size)

        # Set the client size and bind paint event
        self.SetClientSize(window_size)
        self.Bind(wx.EVT_PAINT, self.on_paint)

        # Create an image to mask that is slightly smaller than client area.
        # The mask white defines where image is drawn.
        width = window_size.width - IMAGE_SIZE_REDUCTION
        height = window_size.height - IMAGE_SIZE_REDUCTION
        self.image_bmp = wx.Bitmap(wx.Size(width, height))
        image_dc = wx.MemoryDC(self.image_bmp)
        image_dc.SetBackground(wx.WHITE_BRUSH)
        image_dc.Clear()
        image_dc.SetPen(wx.BLACK_PEN)
        for x in range(0, width, 10):
            image_dc.DrawLine(x, 0, x, height)
        for y in range(0, height, 10):
            image_dc.DrawLine(0, y, width, y)
        #self.image_bmp.SaveFile("image.bmp", wx.BITMAP_TYPE_BMP)
        image_dc.SelectObject(wx.NullBitmap) # release dc to update image_dc

        # Create a mask the same size as image.  Black is mask, White is drawn.
        mask_corners = [(MASK_BORDER, MASK_BORDER),
                        (width-MASK_BORDER, MASK_BORDER),
                        (width-MASK_BORDER+MASK_SKEW, height-MASK_BORDER),
                        (MASK_BORDER+MASK_SKEW, height-MASK_BORDER)]
        self.mask_bmp = wx.Bitmap(self.image_bmp.GetSize(), depth=1)
        mask_dc = wx.MemoryDC(self.mask_bmp)
        gcdc = wx.GCDC(mask_dc)
        gc = gcdc.GetGraphicsContext()
        gc.SetAntialiasMode(wx.ANTIALIAS_NONE) # Needed so mask is B/W on macOS.
        mask_dc.SetBackground(wx.BLACK_BRUSH)
        mask_dc.Clear()
        mask_dc.SetBrush(wx.WHITE_BRUSH)
        mask_dc.DrawPolygon(mask_corners)
        #self.mask_bmp.SaveFile("mask.bmp", wx.BITMAP_TYPE_BMP)
        mask_dc.SelectObject(wx.NullBitmap) # release dc to update mask_bmp
        mask = wx.Mask(self.mask_bmp)

        # Apply the mask to image
        self.image_bmp.SetMask(mask)

    def on_paint(self, _event):
        """ Handle EVT_PAINT events. """

        # Draw the image
        dc = wx.PaintDC(self)
        dc.SetBackground(wx.GREEN_BRUSH)
        dc.Clear()
        dc.DrawBitmap(self.image_bmp, 0, 0, useMask=True)

class MyApp(wx.App):

    def OnInit(self):
        print('sys.version:', sys.version)
        print('wx.version:', wx.version())
        self.frame = MyFrame()
        self.frame.Show()
        self.SetTopWindow(self.frame)
        return True

app = MyApp(0)
app.MainLoop()
```
